pkg/cache/v3/linear.go

```diff
@@ -165,6 +165,41 @@ func (cache *LinearCache) DeleteResource(name string) error {
 	return nil
 }
 
+// SetResources replaces current resources with a new set of resources
+// This function is useful for use cases when DiscoveryRequest#resourceNames is empty.
+// This way watches that follow all resources are triggered only once regardless of how many resources are changed.
+func (cache *LinearCache) SetResources(resources map[string]types.Resource) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	cache.version += 1
+
+	modified := map[string]struct{}{}
+	for name := range cache.resources {
+		if _, found := resources[name]; !found {
+			delete(cache.versionVector, name)
+			modified[name] = struct{}{}
+		}
+	}
+
+	cache.resources = resources
+	for name := range resources {
+		// We assume all resources passed to SetResources are changed.
+		// Otherwise we would have to do proto.Equal on resources which is pretty expensive operation
```

Comment:
There are other solutions to this (like per version resource values like how Delta is modeled), but it's probably fine to leave it as this for now.